diff --git a/zookeeper/test/com/yahoo/zookeeper/test/CreateHammerTest.java b/zookeeper/test/com/yahoo/zookeeper/test/TestHammer.java
similarity index 89%
rename from zookeeper/test/com/yahoo/zookeeper/test/CreateHammerTest.java
rename to zookeeper/test/com/yahoo/zookeeper/test/TestHammer.java
index 0ec65e84b..0661870b5 100644
--- a/zookeeper/test/com/yahoo/zookeeper/test/CreateHammerTest.java
+++ b/zookeeper/test/com/yahoo/zookeeper/test/TestHammer.java
@@ -1,4 +1,3 @@
-package com.yahoo.zookeeper.test;
 
 import java.io.IOException;
 
@@ -8,7 +7,7 @@
 import com.yahoo.zookeeper.ZooDefs.CreateFlags;
 import com.yahoo.zookeeper.ZooDefs.Ids;
 
-public class CreateHammerTest implements VoidCallback {
+public class TestHammer implements VoidCallback {
 
     /**
      * @param args
@@ -30,7 +29,7 @@ public static void main(String[] args) {
                 try {
                     String name = zk.create("/testFile-", new byte[16], Ids.OPEN_ACL_UNSAFE,
                         CreateFlags.EPHEMERAL|CreateFlags.SEQUENCE);
-                    zk.delete(name, -1, new CreateHammerTest(), null);
+                    zk.delete(name, -1, new TestHammer(), null);
                 } catch(Exception e) {
                     i--;
                     e.printStackTrace();
