diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/DataTree.java b/zookeeper/java/src/com/yahoo/zookeeper/server/DataTree.java
index 0f1f1fc9e..530e829d6 100755
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/DataTree.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/DataTree.java
@@ -411,7 +411,7 @@ void killSession(long session) {
                 try {
                     deleteNode(path);
                     ZooLog.logTextTraceMessage("Deleting ephemeral node "
-                            + path + " for session " + session,
+                            + path + " for session " + Long.toHexString(session),
                             ZooLog.SESSION_TRACE_MASK);
                 } catch (KeeperException e) {
                     ZooLog.logException(e);
diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/ZooLog.java b/zookeeper/java/src/com/yahoo/zookeeper/server/ZooLog.java
index 618aa64bc..7e75511bc 100644
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/ZooLog.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/ZooLog.java
@@ -62,7 +62,7 @@ public static void logError(String mess) {
         StackTraceElement ste = re.getStackTrace()[1];
         String location = ZooLog.stackTrace2Location(ste);
         ZooLog.logError(mess, location);
-        ZooLog.logTextTraceMessage(mess + "location: \n" + location,
+        ZooLog.logTextTraceMessage(mess + "location: " + location,
                 textTraceMask);
     }
 
@@ -193,7 +193,7 @@ static private void writeText(FileChannel os, char rp, Request request,
         sb.append(header).append(":").append(rp);
         sb.append(request.toString());
         write(os, sb.toString());
-        write(textTos, "\n\n");
+        write(textTos, "\n");
     }
 
     static private void writeText(FileChannel os, String message,
@@ -206,7 +206,7 @@ static private void writeText(FileChannel os, String message,
                 .append(location).append(" ");
         sb.append(message);
         write(os, sb.toString());
-        write(textTos, "\n\n");
+        write(textTos, "\n");
     }
 
     static private long ROLLOVER_TIME = 24 * 3600 * 1000;
@@ -235,19 +235,19 @@ synchronized private static void checkTextTraceFile() {
             if (requestTraceFile == null) {
                 return;
             }
-            requestTraceFile += "." + year + "." + month + "." + day;
+           String currentTextFile = requestTraceFile +  "." + year + "." + month + "." + day;
 
             try {
-                textTos = new FileOutputStream(requestTraceFile + ".txt", true)
+                textTos = new FileOutputStream(currentTextFile + ".txt", true)
                         .getChannel();
                 textTosCreationTime = time;
-                write(textTos, "\n\n");
+                write(textTos, "\n");
             } catch (IOException e) {
                 ZooLog.logException(e);
                 return;
             }
             ZooLog.logWarn("*********** Traced requests text saved to "
-                    + requestTraceFile + ".txt");
+                    + currentTextFile + ".txt");
         }
     }
 
