diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/FollowerZooKeeperServer.java b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/FollowerZooKeeperServer.java
index f75d8fdd3..da95337c9 100644
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/FollowerZooKeeperServer.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/FollowerZooKeeperServer.java
@@ -54,7 +54,7 @@ class FollowerZooKeeperServer extends ZooKeeperServer {
      */
     FollowerZooKeeperServer(long serverId, File dataDir, File dataLogDir,
             Follower follower) throws IOException {
-        super(dataDir, dataLogDir);
+        super(dataDir, dataLogDir, follower.self.tickTime);
         this.serverId = serverId;
         this.follower = follower;
     }
diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/LeaderZooKeeperServer.java b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/LeaderZooKeeperServer.java
index 075b3d0d7..d2d9b9766 100644
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/LeaderZooKeeperServer.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/LeaderZooKeeperServer.java
@@ -45,7 +45,7 @@ public class LeaderZooKeeperServer extends ZooKeeperServer {
      */
     LeaderZooKeeperServer(long serverId, File dataDir, File dataLogDir,
             Leader leader) throws IOException {
-        super(dataDir, dataLogDir);
+        super(dataDir, dataLogDir, leader.self.tickTime);
         this.serverId = serverId;
         this.leader = leader;
     }
diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java
index cf242b533..2462e3362 100644
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java
@@ -505,7 +505,7 @@ public static void main(String args[]) {
                 self.join();
             } else {
                 int port = clientPort;
-                ZooKeeperServer zk = new ZooKeeperServer(dataDir, dataLogDir);
+                ZooKeeperServer zk = new ZooKeeperServer(dataDir, dataLogDir, tickTime);
                 zk.startup();
                 NIOServerCnxn.Factory t = new NIOServerCnxn.Factory(port);
                 t.setZooKeeperServer(zk);
