diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/PrepRequestProcessor.java b/zookeeper/java/src/com/yahoo/zookeeper/server/PrepRequestProcessor.java
index c821aaaf8..4cf9ab166 100644
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/PrepRequestProcessor.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/PrepRequestProcessor.java
@@ -373,6 +373,12 @@ protected void pRequest(Request request) {
         nextProcessor.processRequest(request);
     }
 
+    /**
+     * 
+     * @param authInfo list of ACL IDs associated with the client connection
+     * @param acl list of ACLs being assigned to the node (create or setACL operation)
+     * @return
+     */
     private boolean fixupACL(ArrayList<Id> authInfo, ArrayList<ACL> acl) {
         if (skipACL) {
             return true;
