diff --git a/zookeeper/c/tests/CppAssertHelper.h b/zookeeper/c/tests/CppAssertHelper.h
index 423b13231..56793cf17 100644
--- a/zookeeper/c/tests/CppAssertHelper.h
+++ b/zookeeper/c/tests/CppAssertHelper.h
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #ifndef CPPASSERTHELPER_H_
 #define CPPASSERTHELPER_H_
 
diff --git a/zookeeper/c/tests/LibCMocks.cc b/zookeeper/c/tests/LibCMocks.cc
index 2f518b776..4503d6f60 100644
--- a/zookeeper/c/tests/LibCMocks.cc
+++ b/zookeeper/c/tests/LibCMocks.cc
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #include <cstdlib>
 #include <cstdarg>
 #include <iostream>
diff --git a/zookeeper/c/tests/LibCMocks.h b/zookeeper/c/tests/LibCMocks.h
index 539768158..e58f95314 100644
--- a/zookeeper/c/tests/LibCMocks.h
+++ b/zookeeper/c/tests/LibCMocks.h
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #ifndef LIBCMOCKS_H_
 #define LIBCMOCKS_H_
 
diff --git a/zookeeper/c/tests/LibCSymTable.cc b/zookeeper/c/tests/LibCSymTable.cc
index 2d7863f80..950a15cbe 100644
--- a/zookeeper/c/tests/LibCSymTable.cc
+++ b/zookeeper/c/tests/LibCSymTable.cc
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #include "LibCSymTable.h" 
 
 #define LOAD_SYM(sym) \
diff --git a/zookeeper/c/tests/LibCSymTable.h b/zookeeper/c/tests/LibCSymTable.h
index 8d176ed35..78b916b6c 100644
--- a/zookeeper/c/tests/LibCSymTable.h
+++ b/zookeeper/c/tests/LibCSymTable.h
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #ifndef LIBCSYMTABLE_H_
 #define LIBCSYMTABLE_H_
 
diff --git a/zookeeper/c/tests/MocksBase.cc b/zookeeper/c/tests/MocksBase.cc
index c33420a53..774e0ffd3 100644
--- a/zookeeper/c/tests/MocksBase.cc
+++ b/zookeeper/c/tests/MocksBase.cc
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #include <cstdlib>
 #include <new>
 
diff --git a/zookeeper/c/tests/MocksBase.h b/zookeeper/c/tests/MocksBase.h
index cfa1121f6..53f62c3a6 100644
--- a/zookeeper/c/tests/MocksBase.h
+++ b/zookeeper/c/tests/MocksBase.h
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #ifndef MOCKSBASE_H_
 #define MOCKSBASE_H_
 
diff --git a/zookeeper/c/tests/PthreadMocks.cc b/zookeeper/c/tests/PthreadMocks.cc
index a6c0c741a..4d256234d 100644
--- a/zookeeper/c/tests/PthreadMocks.cc
+++ b/zookeeper/c/tests/PthreadMocks.cc
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #include "PthreadMocks.h"
 #include "LibCSymTable.h"
 #include "Util.h"
diff --git a/zookeeper/c/tests/PthreadMocks.h b/zookeeper/c/tests/PthreadMocks.h
index 6c22429de..a7363a93d 100644
--- a/zookeeper/c/tests/PthreadMocks.h
+++ b/zookeeper/c/tests/PthreadMocks.h
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #ifndef PTHREADMOCKS_H_
 #define PTHREADMOCKS_H_
 
diff --git a/zookeeper/c/tests/TestDriver.cc b/zookeeper/c/tests/TestDriver.cc
index 796b7d3cb..0361cefc2 100644
--- a/zookeeper/c/tests/TestDriver.cc
+++ b/zookeeper/c/tests/TestDriver.cc
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #include <string>
 #include <cppunit/TestRunner.h>
 #include <cppunit/CompilerOutputter.h>
diff --git a/zookeeper/c/tests/TestOperations.cc b/zookeeper/c/tests/TestOperations.cc
index d30605fbd..1da9f049d 100644
--- a/zookeeper/c/tests/TestOperations.cc
+++ b/zookeeper/c/tests/TestOperations.cc
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #include <cppunit/extensions/HelperMacros.h>
 
 #include <zookeeper.h>
diff --git a/zookeeper/c/tests/TestZookeeperClose.cc b/zookeeper/c/tests/TestZookeeperClose.cc
index 0c9e32883..8929d0356 100644
--- a/zookeeper/c/tests/TestZookeeperClose.cc
+++ b/zookeeper/c/tests/TestZookeeperClose.cc
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #include <cppunit/extensions/HelperMacros.h>
 
 #include <zookeeper.h>
diff --git a/zookeeper/c/tests/TestZookeeperInit.cc b/zookeeper/c/tests/TestZookeeperInit.cc
index 803f3d7be..2e04d450b 100644
--- a/zookeeper/c/tests/TestZookeeperInit.cc
+++ b/zookeeper/c/tests/TestZookeeperInit.cc
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #include <cppunit/extensions/HelperMacros.h>
 
 #include <zookeeper.h>
diff --git a/zookeeper/c/tests/Util.cc b/zookeeper/c/tests/Util.cc
index 8bb3f2440..862afa236 100644
--- a/zookeeper/c/tests/Util.cc
+++ b/zookeeper/c/tests/Util.cc
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #include <zookeeper.h>
 #include "Util.h"
 #include "LibCSymTable.h"
diff --git a/zookeeper/c/tests/Util.h b/zookeeper/c/tests/Util.h
index e85373d77..f61a53bf9 100644
--- a/zookeeper/c/tests/Util.h
+++ b/zookeeper/c/tests/Util.h
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #ifndef UTIL_H_
 #define UTIL_H_
 
diff --git a/zookeeper/c/tests/ZKMocks.cc b/zookeeper/c/tests/ZKMocks.cc
index 0f905b757..10cb13424 100644
--- a/zookeeper/c/tests/ZKMocks.cc
+++ b/zookeeper/c/tests/ZKMocks.cc
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #include <arpa/inet.h>  // for htonl
 #include <memory>
 
diff --git a/zookeeper/c/tests/ZKMocks.h b/zookeeper/c/tests/ZKMocks.h
index 9e7c368ef..970a61264 100644
--- a/zookeeper/c/tests/ZKMocks.h
+++ b/zookeeper/c/tests/ZKMocks.h
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2008, Yahoo! Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 #ifndef ZKMOCKS_H_
 #define ZKMOCKS_H_
 
