diff --git a/zookeeper/java/ChangeLog b/zookeeper/java/ChangeLog
new file mode 100644
index 000000000..08a9b0bde
--- /dev/null
+++ b/zookeeper/java/ChangeLog
@@ -0,0 +1,6 @@
+Release 1.0.0
+2007-11-27 Andrew Kornev <akornev@users.sourceforge.net>
+
+	* Updated the jute compiler to emit int32_t vs int in the generated C code
+	
+	* Changed release numbering scheme to match that of the C client
\ No newline at end of file
