diff --git a/zookeeper/java/src/com/yahoo/zookeeper/ZooKeeper.java b/zookeeper/java/src/com/yahoo/zookeeper/ZooKeeper.java
index f6286d1aa..d8cf8f7da 100755
--- a/zookeeper/java/src/com/yahoo/zookeeper/ZooKeeper.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/ZooKeeper.java
@@ -667,7 +667,6 @@ public void sync(String path, VoidCallback cb, Object ctx){
     	SyncResponse response = new SyncResponse();
     	request.setPath(path);
     	cnxn.queuePacket(h, new ReplyHeader(), request, response, cb, path, ctx);
-    	System.out.println("Sending sync!");
     }
     
     public States getState() {
diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/SessionTrackerImpl.java b/zookeeper/java/src/com/yahoo/zookeeper/server/SessionTrackerImpl.java
index d9d0a4764..2d4d86c74 100644
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/SessionTrackerImpl.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/SessionTrackerImpl.java
@@ -142,7 +142,7 @@ synchronized public void run() {
 
     synchronized public boolean touchSession(long sessionId, int timeout) {
         ZooLog.logTextTraceMessage("SessionTrackerImpl --- Touch session: "
-                + sessionId + " with timeout " + timeout,
+                + Long.toHexString(sessionId) + " with timeout " + timeout,
                 ZooLog.CLIENT_PING_TRACE_MASK);
         Session s = sessionsById.get(sessionId);
         if (s == null) {
@@ -172,7 +172,7 @@ synchronized public void removeSession(long sessionId) {
         Session s = sessionsById.remove(sessionId);
         sessionsWithTimeout.remove(sessionId);
         ZooLog.logTextTraceMessage("SessionTrackerImpl --- Removing "
-                + sessionId, ZooLog.SESSION_TRACE_MASK);
+                + Long.toHexString(sessionId), ZooLog.SESSION_TRACE_MASK);
         if (s != null) {
             sessionSets.get(s.tickTime).sessions.remove(s);
         }
@@ -195,11 +195,11 @@ synchronized public void addSession(long id, int sessionTimeout) {
         if (sessionsById.get(id) == null) {
             Session s = new Session(id, 0);
             sessionsById.put(id, s);
-            ZooLog.logTextTraceMessage("SessionTrackerImpl --- Adding " + id
+            ZooLog.logTextTraceMessage("SessionTrackerImpl --- Adding " + Long.toHexString(id)
                     + " " + sessionTimeout, ZooLog.SESSION_TRACE_MASK);
         } else {
             ZooLog.logTextTraceMessage(
-                    "SessionTrackerImpl --- Existing session " + id + " "
+                    "SessionTrackerImpl --- Existing session " + Long.toHexString(id) + " "
                             + sessionTimeout, ZooLog.SESSION_TRACE_MASK);
         }
         touchSession(id, sessionTimeout);
diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/ZooKeeperServer.java b/zookeeper/java/src/com/yahoo/zookeeper/server/ZooKeeperServer.java
index 2a50b61ce..b63224d8d 100755
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/ZooKeeperServer.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/ZooKeeperServer.java
@@ -347,7 +347,7 @@ public long playLog(InputArchive logStream) throws IOException {
                             ((CreateSessionTxn) txn).getTimeOut());
                     ZooLog.logTextTraceMessage(
                             "playLog --- create session in log: "
-                                    + hdr.getClientId() + " with timeout: "
+                                    + Long.toHexString(hdr.getClientId()) + " with timeout: "
                                     + ((CreateSessionTxn) txn).getTimeOut(),
                             ZooLog.SESSION_TRACE_MASK);
                     // give dataTree a chance to sync its lastProcessedZxid
@@ -357,7 +357,7 @@ public long playLog(InputArchive logStream) throws IOException {
                     sessionsWithTimeouts.remove(hdr.getClientId());
                     ZooLog.logTextTraceMessage(
                             "playLog --- close session in log: "
-                                    + hdr.getClientId(),
+                                    + Long.toHexString(hdr.getClientId()),
                             ZooLog.SESSION_TRACE_MASK);
                     dataTree.processTxn(hdr, txn);
                     break;
diff --git a/zookeeper/test/com/yahoo/zookeeper/test/AsyncTest.java b/zookeeper/test/com/yahoo/zookeeper/test/AsyncTest.java
index 32e0290de..cf37bec0b 100644
--- a/zookeeper/test/com/yahoo/zookeeper/test/AsyncTest.java
+++ b/zookeeper/test/com/yahoo/zookeeper/test/AsyncTest.java
@@ -45,7 +45,7 @@ protected void restart() throws Exception {
     @After
     protected void tearDown() throws Exception {
         qt.tearDown();
-    	ZooLog.logError("Clent test shutdown");
+    	ZooLog.logError("Client test shutdown");
         if (f != null) {
             f.shutdown();
         }
