diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/FollowerHandler.java b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/FollowerHandler.java
index 3cadd76ca..5bc083cef 100644
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/FollowerHandler.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/FollowerHandler.java
@@ -264,7 +264,7 @@ public void run() {
                     int type = bb.getInt();
                     bb = bb.slice();
                     leader.zk.submitRequest(null, sessionId, type, cxid, bb,
-                            null);
+                            qp.getAuthinfo());
                     break;
                 default:
                 }
