diff --git a/zookeeper/java/src/com/yahoo/jute/compiler/CGenerator.java b/zookeeper/java/src/com/yahoo/jute/compiler/CGenerator.java
index ca331039f..d971572ec 100644
--- a/zookeeper/java/src/com/yahoo/jute/compiler/CGenerator.java
+++ b/zookeeper/java/src/com/yahoo/jute/compiler/CGenerator.java
@@ -64,6 +64,9 @@ void genCode() throws IOException {
             JFile f = (JFile) i.next();
             h.write("#include \""+f.getName()+".h\"\n");
         }
+        // required for compilation from C++
+        h.write("\n#ifdef __cplusplus\nextern \"C\" {\n#endif\n\n");
+
         c.write("#include <stdlib.h>\n"); // need it for calloc() & free()
         c.write("#include \""+mName+".h\"\n\n");
         
@@ -72,6 +75,7 @@ void genCode() throws IOException {
             jr.genCCode(h, c);
         }
         
+        h.write("\n#ifdef __cplusplus\n}\n#endif\n\n");
         h.write("#endif //"+mName.toUpperCase().replace('.','_')+"__\n");
         
         h.close();
