diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java
index 2462e3362..cc0426a74 100644
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java
@@ -103,7 +103,6 @@ public QuorumServer(long id, InetSocketAddress addr) {
     public enum ServerState {
         LOOKING, FOLLOWING, LEADING;
     }
-
     /**
      * The servers that make up the cluster
      */
@@ -114,6 +113,14 @@ public enum ServerState {
      */
     long myid;
 
+
+    /**
+     * get the id of this quorum peer.
+     */
+    public long getId() {
+    	return myid;
+    }
+
     /**
      * This is who I think the leader currently is.
      */
