diff --git a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java
index 2093740fd..cf242b533 100644
--- a/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java
+++ b/zookeeper/java/src/com/yahoo/zookeeper/server/quorum/QuorumPeer.java
@@ -474,11 +474,11 @@ public static void main(String args[]) {
                 ZooLog.logError("tickTime is not set");
                 System.exit(2);
             }
-            if (initLimit == 0) {
+            if (servers.size() > 1 && initLimit == 0) {
                 ZooLog.logError("initLimit is not set");
                 System.exit(2);
             }
-            if (syncLimit == 0) {
+            if (servers.size() > 1 && syncLimit == 0) {
                 ZooLog.logError("syncLimit is not set");
                 System.exit(2);
             }
@@ -517,4 +517,4 @@ public static void main(String args[]) {
         }
         System.exit(2);
     }
-}
\ No newline at end of file
+}
